### DISCUSSION TITLE
Stop mutation process

### DISCUSSION THREAD

**Discussion by BlackGad** [stryker-mutator/stryker-net#2150] [Ideas]
Is it possible to extend stryker flow with additional pre conditions? For example if any or specific unit test failed after initial test run?

Currently there is only one condition (50% of tests are failed).

**Comment by rouke-broersma** [comment]
I don't think we need to make this unnecessarily complex within stryker. if you have very specific preconditions, make them preconditions in whatever automation you use. We already provide testcase filter support, and we could provide support for not allowing any failing unit tests as well (binary option, allow failing unit tests default true, but no option to specify percentage of failing unit tests). If you would like support for not allowing any failing unit tests please create a feature request.

**Comment by psfinaki** [comment]
Hey there, a bit late to the party but first of all thanks for implementing this, second just wanted to link [this discussion](https://github.com/stryker-mutator/stryker-net/discussions/1898) on the same topic I started back in a day.

My opinion is that the implemented solution can probably even deprecate the 50%-breaking-tests approach. That will simplify the code. If so, I can take care of removing it (removing code is always a pleasure :D). But well surely it can stay also.

**Reply by rouke-broersma** [reply]
Then we could have gone back to the previous implementation where we always fail on failing unit tests. That's how Stryker worked for the longest time. We changed that because now that we have coverage information it's no longer necessary to fail fast. 

A non working unit test is essentially the same as a missing unit test imo.

Or are you suggesting removing the, admittedly randomish chosen, percentage of 50 percent and rather always running if there is at least one functioning unit test?

**Reply by psfinaki** [reply]
Yes, the latter, I suggest removing the 50% tests approach. So that the newly added flag just replaces the previous behavior. You either care about failing tests or not (and take responsibility then).

In accord with my original suggestion, I'd even go as far as setting up the new flag to `true` by default because I imagine that's what most of the people would want. But that will be technically a breaking change.

**Reply by rouke-broersma** [reply]
I don't think we will want to change the default back to fail by default. At least for now we've had one user request for this behavior, and your questioning why it is this way. This to me is not enough feedback to suggest we need to change the default. 

@dupdob @richardwerkman what do you guys think about changing the 50 percent failure threshold to a any test succeeded threshold?

**Reply by BlackGad** [reply]
I think Stryker not about tests success but about code run. Even failed test run some code that can cover something. I think you need remove default failing even if all tests failed. With new option users will decide what they want. Brief code analyze or clean code run.

**Reply by rouke-broersma** [reply]
No that's not correct I'm afraid. Stryker can only discover if a mutant is covered by a test if a test fails for that mutant. So that means that any test that is failing before we have mutated must be excluded from mutation testing, it cannot give us any results.

**Reply by dupdob** [reply]
I don't remember the source of this feature (discussion or issue), but it was added because users complained not being able to use Stryker when they had **some** flaky tests. This typically happened with external dependencies (for tests). 

The logic is that we need to tolerate 'some failing tests', including for small test base, but also abort running Stryker if there was clearly a problem with a code base. 

After careful discussion it was decided with @richardwerkman (I think) that 50% was as good as any other number.

At the time, we said we would make this threshold configurable if people asked for it. 

Personally, I think Stryker should tolerate failing tests (there are other pipeline gates to prevent this) while avoiding useless runs, such as when there is only one test failing passing. 


@BlackGad : actually, failing tests have no value for Stryker, as we don't know is the test fails 'naturally' or due to the mutation we are testing. 
Adding support for failing was quite an effort, it is trickier than it looks at first sight

**Reply by BlackGad** [reply]
> No that's not correct I'm afraid. Stryker can only discover if a mutant is covered by a test if a test fails for that mutant. So that means that any test that is failing before we have mutated must be excluded from mutation testing, it cannot give us any results.

Ah base line. Got it :)

**Reply by psfinaki** [reply]
Right, flaky tests, yeah didn't think about them. Well anyway, even this discussion shows there's not much consensus here so probably we should keep things as they are. Although I guess these two rules can be really just merged into one, having the allowed failing tests percentage specified in the config.

**Reply by richardwerkman** [reply]
> After careful discussion it was decided with richardwerkman (I think) that 50% was as good as any other number.

Yes, we did. It didn't make sense to run stryker with mostly failing tests as the results will for sure be invalid. We could remove the threshold as suggested but I'm not sure what we will gain with that. But now that there is a possibility to break on failing tests I'd be fine with removing the threshold completely if it's what some people want.

But I'm thinking how we would handle 0 succeeding tests. Would we fail or just return a report with a 0 mutation score?

**Reply by psfinaki** [reply]
IMO, 0 succeeding tests is a very miserable situation and any tool that notices it should cry with pain. Another thing is that 0 mutation score is usually interpreted as "my code is not tested at all" which would not really be the case here.

**Reply by rouke-broersma** [reply]
I prefer to keep it as it is. We decide what percentage we are comfortable with and can change it at any time if we feel that 50 percent is not correct. If users don't like this behavior they can disable it completely. I don't see much reason to support a custom threshold. I do think we should document this behavior better.
